state/relation.go

```diff
@@ -110,43 +111,117 @@ func (r *Relation) Status() (status.StatusInfo, error) {
 
 // SetStatus sets the status of the relation.
 func (r *Relation) SetStatus(statusInfo status.StatusInfo) error {
```

Comment:
This feels really inappropriate for setting the status. I think it's down to making use of SetStatus for re-joining a relation, which also feels really inappropriate (did I approve that? if so, sorry)

Status is for reporting how things are, not for how we want them to be. When you re-join a relation, you shouldn't be updating the status directly. Instead, you should be setting a flag on the relation, saying you want it to be re-joined. If it's acceptable, and happens immediately, it can *also* update the status; if it happens (if we change it to happen) asynchronously, then status would be updated later when it is actually joined.

Reply:
When a relation is first created, it's status starts out as "joined". ie it is "alive" and "joined". Semantically, the resume operation is putting the relation back in the "joined" state which is intended to be how things are. A relation is either "joined" or "suspended". I started out using the term "active" (so "active" or "suspended") but the policy of not reusing status names means I had to think of a new name. Perhaps there's some confusion around the semantics due to the names used.

Reply:
I don't really care about the names.

What I'm trying to get at is that the client has no business setting the relation status directly. It *might* be set as a side-effect of doing something else, if that something else is a completely synchronous operation. Status is purely for reporting the current status of something, so it does not make sense to be checking permissions in there.

The user should be setting a property on the relation, e.g. "suspended". Setting that should be predicated on permissions.